### PR TITLE
Implements icon size caching to avoid unnecessary load in updatehealth

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Technically an improved port of https://github.com/DaedalusDock/daedalusdock/pull/651, instead of only storing height over 32 pixels for HUDs we store both pure height and width for the sake of cutting down on icon operation spam (which is pretty costly). Should save us a significant amount of time, cuts down update_health_hud times by 45% and total update_health by 30% which is pretty good for a somewhat hot proc.

## Why It's Good For The Game

Our health HUDs constantly fetch icons ***twice*** every update_health, jesus.

## Changelog
:cl: SmArtKar, Kapu
code: Implemented caching for icon sizes which should significantly improve mob health performance due to HUDs constantly fetching icons
/:cl:

### CODE REVIEW

File: code/game/data_huds.dm [hunk 1]
```diff
@@ -579,3 +556,28 @@ Diagnostic HUDs!
...
+	return GLOB.icon_size_cache[icon][2]
+
+/atom/proc/cache_icon_size()
+	var/icon/sample_icon
+	if (isfile(icon))
+		sample_icon = icon(icon, icon_state, dir)
+	else
+		sample_icon = icon
+	// If icon is a file, we are guaranteed to always have a consistent sprite size for all objects from that icon
+	GLOB.icon_size_cache[isfile(icon) ? icon : sample_icon] = list(sample_icon.Width(), sample_icon.Height())
```
Comment (LemonInTheDark) [inline code/game/data_huds.dm:583]: worth noting icon_states calls are at least partially internally cached, it's not 100% that this is worthwhile. likely is but wanted to note

File: code/game/data_huds.dm [hunk 2]
```diff
@@ -579,3 +556,28 @@ Diagnostic HUDs!
...
+		cache_icon_size()
+	return GLOB.icon_size_cache[icon][1]
+
+/atom/proc/get_cached_height()
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if (isnull(icon))
+		return null
+	if (!GLOB.icon_size_cache[icon])
+		cache_icon_size()
+	return GLOB.icon_size_cache[icon][2]
```
Comment (LemonInTheDark) [inline code/game/data_huds.dm:574]: use defines

File: code/game/atom/atom_appearance.dm [hunk 3]
```diff
@@ -1,3 +1,6 @@
+/// Assoc list of icon file/icon to width/height of the icon, used to cut down on RSC operations
+GLOBAL_LIST_INIT(icon_size_cache, list())
```
Comment (LemonInTheDark) [inline code/game/atom/atom_appearance.dm:2]: why here if the relevant procs are not?

**Comment by LemonInTheDark** [comment]
needs profiling

File: code/game/data_huds.dm (hunk 1 above)
Comment (SmArtKar) [reply code/game/data_huds.dm:583]: having some issues with tracy, will post profiler results slightly later

File: code/game/data_huds.dm [hunk 4]
```diff
@@ -579,3 +556,28 @@ Diagnostic HUDs!
...
+
+/atom/proc/cache_icon_size()
```
[the remark this answers is not in the document]
Comment (SmArtKar) [reply code/game/data_huds.dm:576]: what should this be tied to instead?

**Comment by SmArtKar** [comment]
~45% improvement, tested on 10 humans after tossing an ezclean grenade in both cases. This (yellow) is with the optimization, external (red) is without - pretty significant improvement.

<details>

<summary>Profile screens</summary>

tracy-profiler_mREsvFOSdB
tracy-profiler_9ogqXDL3lO

</details>

File: code/game/data_huds.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/data_huds.dm]: global proc that is called into by a wrapper proc on atom

**Comment by SmArtKar** [comment]
Turns out we already had a better helper for this, except it wasn't used anywhere outside of fishing, microwaves and a few niche cases. Used it instead + hardcoded values for hottest users (humans) per @AnturK's suggestion, got roughly the same results in the profiler.

**Pull request merged**